### PR TITLE
Linux crash handler

### PR DESCRIPTION
**Pull request by daleglass**
Initial implementation of Linux crashpad support.

This builds and uses a Linux crashpad binary archive I generated. This needs testing. I'm not clear on that it actually uploads crashes all the way, but it certainly seems to be doing something and not producing any fatal errors.

It generates a warning on build:

```
CMake Warning (dev) at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:272 (message):
  The package name passed to `find_package_handle_standard_args` (CRASHPAD)
  does not match the name of the calling package (Crashpad).  This can lead
  to problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  cmake/modules/FindCrashpad.cmake:41 (find_package_handle_standard_args)
  /home/dale/Vircadia/vcpkg/5c613fb5/scripts/buildsystems/vcpkg.cmake:262 (_find_package)
  cmake/macros/AddCrashpad.cmake:31 (find_package)
  interface/CMakeLists.txt:242 (add_crashpad)
This warning is for project developers.  Use -Wno-dev to suppress it.
```

I could use somebody knowledgeable in cmake to fix this one.

### CODE REVIEW

**Comment by daleglass** [comment]
The scripts to generate the crashpad binaries can be found at https://github.com/daleglass/crashpad-builder

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR365/ProjectAthena-Alpha-PR365-fdaa4d0_9fcd8ef6.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR365/ProjectAthena-Alpha-PR365-fdaa4d0.tar_9fcd8ef6.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR365/ProjectAthena-Alpha-PR365-fdaa4d0.tar_9fcd8ef6.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR365/ProjectAthena-Alpha-PR365-fdaa4d0_9fcd8ef6.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR365/ProjectAthena-Alpha-PR365-fdaa4d0_9fcd8ef6.exe

**Comment by daleglass** [comment]
On crashing the interface, this happens:

```
[05/18 22:06:55] [DEBUG] [hifi.shared] About to make a pure virtual call
pure virtual method called
terminate called without an active exception
[1184956:1184956:20200518,220655.855942:ERROR process_memory_range.cc:86] read out of range
[1184956:1184956:20200518,220655.856011:ERROR elf_image_reader.cc:594] missing nul-terminator
[1184956:1184956:20200518,220655.863951:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864459:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864493:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864523:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864554:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864585:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864615:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864696:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864726:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864755:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864834:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864865:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.864902:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865229:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865259:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865286:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865318:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865418:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865470:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865500:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865529:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865611:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865661:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865715:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865767:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865817:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865869:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865920:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.865979:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.866030:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.866079:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184956:20200518,220655.866126:ERROR elf_dynamic_array_reader.h:61] tag not found
[1184956:1184957:20200518,220655.899952:ERROR http_transport_socket.cc:163] SSL_connect
[1184956:1184957:20200518,220655.899963:ERROR http_transport_socket.cc:565] SSLStream Initialize
zsh: abort (core dumped)  interface/interface --allowMultipleInstances
```

Dumping packets indicates that a SSL connection is established and quickly fails after 7 packets. So far no clue of what's going wrong.

**Comment by daleglass** [comment]
I captured the HTTP request my interface tries to submit here:

http://hifi.daleglass.net/files/crash_report

It can be sent to a Sentry server like this (the `cat` is there so that `nc` doesn't immediately disconnect, and sticks around for receiving the response):

`cat crash_report /dev/stdin | nc sentry.vircadia.dev 80`

or:

`cat crash_report /dev/stdin | nc --ssl sentry.vircadia.dev 443` for SSL


Apparently the Windows client works, so perhaps there's something wrong in how the crash is submitted.

**Comment by JulianGro** [comment]
So when I submit stuff to sentry from the weblate server, it takes a differnet kind of url. https://id@your.sentry.example.com/
It takes an ID in front of the url. This is the kind of url that django used to submit to sentry.

Not sure if it works the same here, but I thought I should tell you about this.

**Comment by JulianGro** [comment]
On port 80 I actually get a response. I get a 404 not found with a webpage.
The second command for port 443 does not work for me. `nc` compains about `nc: invalid option -- '-'` is I use your command. And if I remove one of the `-` I get `nc: getaddrinfo: Name or service not known` while no connections are being made.

Keep in mind that you can use tools like Wireshark to see what actually happens.

**Comment by daleglass** [comment]
@JulianGro A very annoying thing about netcat is that there are multiple implementations floating around, with incompatible arguments.

File: cmake/macros/AddCrashpad.cmake [hunk 1]
```diff
@@ -23,12 +23,12 @@ macro(add_crashpad)
...
   endif()
 
-  if ((WIN32 OR APPLE) AND USE_CRASHPAD)
+  if (USE_CRASHPAD)
     get_property(CRASHPAD_CHECKED GLOBAL PROPERTY CHECKED_FOR_CRASHPAD_ONCE)
     if (NOT CRASHPAD_CHECKED)
 
       add_dependency_external_projects(crashpad)
-      find_package(crashpad REQUIRED)
+      find_package(Crashpad REQUIRED)
```
Comment (JulianGro) [inline cmake/macros/AddCrashpad.cmake:31]: For the cmake warning: Is there any reason you changed "crashpad" to "Crashpad"? `Fetch is getting "crashpad" https://chromium.googlesource.com/chromium/tools/depot_tools.git/+/refs/heads/master/fetch_configs/crashpad.py and I don't see any "Crashpad" in the files for crashpad either.

**Comment by JulianGro** [comment]
Comparing this to the way the crash handler uploads stuff on Windows, it seems pretty obvious what is wrong. I just don't know why it is doing that.
On Linux it is sending to `/api/2/minidump/?sentry_key=dbf6000201a442e0b5462b63e18ee6b1&guid=69d358fb-261f-437b-9ed0-c5be10733f38` via http 1.0 which returns a 404.
On Windows it is sending to `/api/2/minidump/?sentry_key=dbf6000201a442e0b5462b63e18ee6b1` via http 1.1.
So Sentry is either not accepting http 1.0 or, what my bet is, Sentry isn't accepting the guid query.

**Comment by daleglass** [comment]
I think it's the Host header. Without that, there's no virtual host support.

So either we need to poke at crashpad and make it send the header, or we need to run sentry as a default virtual host to ensure it receives such crash dumps.

**Comment by daleglass** [comment]
Okay, @JulianGro found out that Crashpad just started using libcurl, which hopefully will work better. I updated my docker builder and this PR, and already uploaded an updated binary package.

Also the PR has been rebased against the latest master.

My attempt at testing doesn't seem to have worked yet, but @JulianGro did seem to manage to submit something.

**Comment by JulianGro** [comment]
Little update:
I am currently having two main issues with crash reports. One is that crashpad will crash if it is inside an AppImage for whatever reason. The same build will work fine if I extract the AppImage.
The other issue is that even after uploading debug files to sentry, it still doesn't give us any usable debug information. https://gist.github.com/JulianGro/1a111933ffa7e41e30d0bacc10305f9e While I missed two system libraries on the debug file upload, the main problem I see here is that sentry doesn't know what binary half of the stuff even belongs to. Personally I have never worked on anything like this before though, so I might just be missing some basic requirement.

**Comment by digisomni** [comment]
1 year anniversary!

**Comment by digisomni** [comment]
At dev meeting:

> With Windows, Developer > Crash menu ... some crashes are reported and others aren't reported ... So perhaps for Linux testing, find which crashes are successfully reported by Windows in the hope that these are more likely to be reported by Linux.

**Comment by JulianGro** [comment]
The reporting itself works, the problem is that Sentry isn't able to extract any meaningful information from them.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Interface-PR365-6bfa5ef-_43e90992.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-6bfa5ef-_43e90992.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-6bfa5ef-.tar_43e90992.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-6bfa5ef-.tar_43e90992.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-6bfa5ef-_43e90992.sh

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-6bfa5ef_43e90992.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Quest-Alpha-PR365-6bfa5ef_43e90992.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-6bfa5ef-_43e90992.exe

**Comment by daleglass** [comment]
Bloody hell. I just found this in the crashpad documentation:

```
--no-rate-limit
Do not rate limit the upload of crash reports. By default uploads are throttled to one per hour. Using this option disables that behavior, and Crashpad will attempt to upload all captured reports.
```

Well, that could explain a fair amount of trouble we've had testing this thing.

**Comment by daleglass** [comment]
Okay, updated to the latest crash reporter code. Reporting seems to work. There's a few things that need addressing, but I think in successive PRs:

* Generate debug info. This should be easy, and could be built into cmake. The issue is then how to arrange uploading that to Sentry.
* Some improvements to the crash reporter may be in order. Currently, disabling the crash reporter disables the entire mechanism. What I think should be done instead is just disabling uploading. That way, once reporting is enabled, past reports could be uploaded. This needs a note in the UI  of course.
* There exists the possibility of including some logs in the crash report. Unfortunately crashpad lacks a way to attach large amounts of arbitrary data, so this would be hackish.
* Perhaps Sentry has some sort of API we could use to add additional info like screenshots and user text to reports

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Interface-PR365-a4db869-_12db4709.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-a4db869-.tar_12db4709.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-a4db869-_12db4709.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-a4db869-.tar_12db4709.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-a4db869-_12db4709.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-a4db869-_12db4709.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Quest-Alpha-PR365-a4db869_12db4709.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-a4db869_12db4709.apk

**Comment by JulianGro** [comment]
> * Generate debug info. This should be easy, and could be built into cmake. The issue is then how to arrange uploading that to Sentry.

The actual debug info isn't a problem. I can just upload that as I have done in the past. The problem we had last time was that the reports we got on sentry were pretty much useless.

**Comment by daleglass** [comment]
Okay, I think I see it!

You have to upload debug symbols before the crash report comes in, and uploading it afterwards doesn't seem to fix the issue. I just tested a new crash after uploading symbols for interface, and that seems to have worked.

**Comment by daleglass** [comment]
@JulianGro I only uploaded symbols for interface, but it looks good to me. The test crash is 838b52f827164e8c9ef598ca7a5af664, std::__atomic_base<T>::load SIGABRT.

**Comment by JulianGro** [comment]
I guess I will generate a test AppImage then.

**Comment by JulianGro** [comment]
The AppImage still crashes if it includes crashpad:
```
[06/29 21:37:30] [DEBUG] [hifi.shared] Settings file: "/home/motofckr9k/.config/Vircadia - dev/Interface.json"
[06/29 21:37:30] [DEBUG] [default] UserActivityLogger is enabled: true
[06/29 21:37:30] [DEBUG] [default] Crash handler logger is enabled: true
[06/29 21:37:30] [DEBUG] [hifi.networking] Final machine fingerprint: "{99ad601f-c559-4a84-4a38-d1d0329f8762}"
[17216:17216:20210629,213730.502233:FATAL CrashHandler_Crashpad.cpp:343] Assertion failed: interfaceDir.exists(CRASHPAD_HANDLER_NAME). 

Thread 1 "AppRun" received signal SIGTRAP, Trace/breakpoint trap.
0x000055555619bbd9 in logging::LogMessage::~LogMessage() ()
(gdb) bt
#0  0x000055555619bbd9 in logging::LogMessage::~LogMessage() ()
#1  0x0000555555c147c2 in startCrashHandler (appPath="/media/motofckr9k/sambawww/m/Vircadia-x86_64_crash_handler_PR365_12db470.AppImage") at /home/motofckr9k/Vircadia/source/interface/src/CrashHandler_Crashpad.cpp:343
#2  0x0000555555e246b3 in main (argc=1, argv=0x7fffffffdb58) at /home/motofckr9k/Vircadia/source/interface/src/main.cpp:226
(gdb) 
```

Here is the AppImage: https://appimage.moto9000.moe/experimental/PR365/Vircadia-x86_64_crash_handler_PR365_12db470.AppImage
This time it also crashes if I extract the AppImage.

**Comment by daleglass** [comment]
Found the problem. You need to include the `crashpad_handler` binary in the AppImage. It should be generated in the `interface/` directory, right alongside the `interface` binary.

That probably would be an improvement to vircadia-builder

**Comment by JulianGro** [comment]
`crashpad_handler` is included in the AppImage under `/vircadia/interface/crashpad_handler`

**Comment by daleglass** [comment]
Discord discussion: interface is looking for `crashpad_handler` in `/`. It works fine on a standalone build, but something seems to break in the AppImage setup. Should be an easy fix.

**Comment by daleglass** [comment]
Pushed a commit to deal with AppImage issues:

* Extensive logging
* Alternate way of finding the application's path on Linux, which should work under AppImage
* Don't abort on startup if we fail to find the `crash_handler` binary.

The debug logging from the crash handler initialization isn't visible by default, run with this to get the debug output:

    QT_LOGGING_RULES='vircadia.crash_handler.debug=true'

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a-.tar_c8b41b8e.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a-.tar_c8b41b8e.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a-_c8b41b8e.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Interface-PR365-b82868a-_c8b41b8e.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Quest-Alpha-PR365-b82868a_c8b41b8e.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a_c8b41b8e.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a-_c8b41b8e.exe

**Comment by JulianGro** [comment]
Update AppImage here: http://appimage.moto9000.moe/experimental/PR365/Vircadia-x86_64_crash-handler_PR365_c8b41b8ea6_build2.AppImage
Works fine on my end.
```
[07/01 04:58:31] [DEBUG] [vircadia.crash_handler] Looking at "/proc/17924/exe"
[07/01 04:58:31] [DEBUG] [vircadia.crash_handler] exe symlink points at QFileInfo(/tmp/.mount_VircadLogFSu/vircadia/interface/interface)
[07/01 04:58:31] [DEBUG] [vircadia.crash_handler] Locating own directory by platform-specific method
[07/01 04:58:31] [DEBUG] [vircadia.crash_handler] Crashpad handler found at "/tmp/.mount_VircadLogFSu/vircadia/interface/crashpad_handler"
[07/01 04:58:31] [DEBUG] [vircadia.crash_handler] Opening crashpad database "/home/motofckr9k/.local/share/Vircadia - dev/Interface/crashpad-db"
[07/01 04:58:31] [INFO] [vircadia.crash_handler] Crashpad initialized
[07/01 04:58:31] [DEBUG] [default] Crash handler started: true
```

Sentry *does* think that this is "PR878_43e909926cb211c629ff59fc9cb12b9e64901e19". I guess it must be using the backtrace token for the info when there is no release otherwise specified?

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a-_c8b41b8e.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a-.tar_c8b41b8e.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a-.tar_c8b41b8e.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a-_c8b41b8e.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Interface-PR365-b82868a-_c8b41b8e.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Quest-Alpha-PR365-b82868a_c8b41b8e.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a_c8b41b8e.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-b82868a-_c8b41b8e.exe

**Review by JulianGro — APPROVED** [review]
Code looks good to me.
I feel like the Windows build should see a quick test.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-fa62688-.tar_1b565794.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-fa62688-_1b565794.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-fa62688-.tar_1b565794.Z

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Interface-PR365-fa62688-_1b565794.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-fa62688-_1b565794.exe

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-fa62688-_1b565794.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Quest-Alpha-PR365-fa62688_1b565794.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-fa62688_1b565794.apk

File: interface/src/CrashHandler_Crashpad.cpp [hunk 2]
```diff
@@ -338,11 +374,29 @@ bool startCrashHandler(std::string appPath) {
     const auto crashpadDbPath = crashpadDbDir.toStdString() + "/" + crashpadDbName;
 
     // Locate Crashpad handler
-    const QFileInfo interfaceBinary{ QString::fromStdString(appPath) };
-    const QDir interfaceDir = interfaceBinary.dir();
-    assert(interfaceDir.exists(CRASHPAD_HANDLER_NAME));
+    QString binaryDir = findBinaryDir();
+    QDir interfaceDir;
+
+    if ( !binaryDir.isEmpty() ) {
```
Comment (ctrlaltdavid) [inline interface/src/CrashHandler_Crashpad.cpp:380]:
```suggestion
    if (!binaryDir.isEmpty()) {
```

File: interface/src/CrashHandler_Crashpad.cpp [hunk 3]
```diff
@@ -311,8 +317,38 @@ void checkUnhandledExceptionHook() {
...
+
+#ifdef Q_OS_LINUX
+    // Find outselves by looking at /proc/<PID>/exe
+    pid_t ourPid = getpid();
+    QString exeLink = QString("/proc/%1/exe").arg(ourPid);
+    qCDebug(crash_handler) << "Looking at" << exeLink;
+
+    QFileInfo exeLinkInfo(exeLink);
+    if (exeLinkInfo.isSymLink()) {
+        QFileInfo exeInfo( exeLinkInfo.symLinkTarget() );
```
Comment (ctrlaltdavid) [inline interface/src/CrashHandler_Crashpad.cpp:337]:
```suggestion
        QFileInfo exeInfo(exeLinkInfo.symLinkTarget());
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-d979597-.tar_30f43e9d.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-d979597-_30f43e9d.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-d979597-.tar_30f43e9d.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-d979597-_30f43e9d.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-d979597_30f43e9d.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Quest-Alpha-PR365-d979597_30f43e9d.apk

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by digisomni** [comment]
The PR build of this does not start or contain crashpad handler on Windows. I'll have to build the PR locally to confirm.

**Comment by daleglass** [comment]
Oops, I just realized this needs some more fixing. Working on it.

**Comment by JulianGro** [comment]
Two reviewers and we both didn't catch that ;P

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-002c0b1-.tar_c0a5868b.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-002c0b1-_c0a5868b.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-002c0b1-.tar_c0a5868b.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-002c0b1-_c0a5868b.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Quest-Alpha-PR365-002c0b1_c0a5868b.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-002c0b1_c0a5868b.apk

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-34a881b-.tar_c0a5868b.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-34a881b-.tar_c0a5868b.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-34a881b-_c0a5868b.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Interface-PR365-34a881b-_c0a5868b.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-34a881b-_c0a5868b.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Quest-Alpha-PR365-34a881b_c0a5868b.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-34a881b_c0a5868b.apk

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-34a881b-_c0a5868b.dmg

**Comment by digisomni** [comment]
Did something change that causes crashpad_handler.exe not to be built/packaged on a PR build? My other PR builds from recent times also don't have it in the folder. Either way, this PR doesn't have crashpad_handler.exe on Windows for some reason.

**Comment by JulianGro** [comment]
I can confirm that this and other PRs don't contain the crash handler on Windows.
We know that this PR works on Linux so should we just give it a green light and fix Windows PR builds in a separate PR?

**Comment by digisomni** [comment]
So long as we're sure it's not this PR causing that but so far it doesn't appear to be this PR's problem.

**Comment by digisomni** [comment]
I need to do a release build then, at least to confirm.

**Comment by daleglass** [comment]
Please try running while setting `QT_LOGGING_RULES='vircadia.crash_handler.debug=true'`. This will produce more debug output.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-7d03226-_e3004af6.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-7d03226-.tar_e3004af6.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-7d03226-.tar_e3004af6.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-7d03226-_e3004af6.exe

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Interface-PR365-7d03226-_e3004af6.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-7d03226_e3004af6.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-Quest-Alpha-PR365-7d03226_e3004af6.apk

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR365/Vircadia-PR365-7d03226-_e3004af6.dmg

**Pull request merged**